File: sdks/rust/conformance-host/src/schema_subset.rs
//! The JSON Schema keywords this adapter can enforce.
//!
//! Fixture `inputSchema` documents are raw JSON, so something has to check
//! invocation input against them. Rather than pull a full JSON Schema
//! implementation into a test binary, this module covers the keywords the
//! corpus actually uses and refuses, at registration time, any schema that
//! needs more. A fixture that would otherwise pass because a keyword was
//! silently ignored fails the run instead.

use serde_json::Value;
use tesseron::ValidationIssue;

const SUPPORTED_KEYWORDS: [&str; 10] = [
    "$schema",
    "additionalProperties",
    "const",
    "default",
    "description",
    "enum",
    "items",
    "properties",
    "required",
    "type",
];

/// Rejects a schema using a keyword this module cannot enforce.
pub fn assert_enforceable(schema: &Value) -> Result<(), String> {
    let Value::Object(members) = schema else {
        return Err(format!("a schema must be a JSON object, got {schema}"));
    };
    for keyword in members.keys() {
        if !SUPPORTED_KEYWORDS.contains(&keyword.as_str()) {
            return Err(format!(
                "this adapter cannot enforce the JSON Schema keyword {keyword:?}"
            ));
        }
    }
    if let Some(Value::Object(properties)) = members.get("properties") {
        for property in properties.values() {
            assert_enforceable(property)?;
        }
    }
    if let Some(items) = members.get("items") {
        assert_enforceable(items)?;
    }
    Ok(())
}

/// Reports every way `value` fails `schema`.
pub fn check(schema: &Value, value: &Value) -> Result<(), Vec<ValidationIssue>> {
    let mut issues = Vec::new();
    collect(schema, value, &mut Vec::new(), &mut issues);
    if issues.is_empty() {
        Ok(())
    } else {
        Err(issues)
    }
}

fn collect(
    schema: &Value,
    value: &Value,
    path: &mut Vec<String>,
    issues: &mut Vec<ValidationIssue>,
) {
    let Value::Object(members) = schema else {
        return;
    };

    if let Some(expected) = members.get("type") {
        if !matches_type(expected, value) {
            issues.push(ValidationIssue::new(
                format!("expected type {expected}, got {}", type_name(value)),
                path.clone(),
            ));
            return;
        }
    }
    if let Some(Value::Array(allowed)) = members.get("enum") {
        if !allowed.contains(value) {
            issues.push(ValidationIssue::new(
                format!("expected one of {}", Value::Array(allowed.clone())),
                path.clone(),
            ));
        }
    }
    if let Some(expected) = members.get("const") {
        if expected != value {
            issues.push(ValidationIssue::new(
                format!("expected {expected}"),
                path.clone(),
            ));
        }
    }

    if let Value::Object(fields) = value {
        if let Some(Value::Array(required)) = members.get("required") {
            for name in required {
                if let Value::String(name) = name {
                    if !fields.contains_key(name) {
                        path.push(name.clone());
                        issues.push(ValidationIssue::new(
                            "required property is missing",
                            path.clone(),
                        ));
                        path.pop();
                    }
                }
            }
        }
        if let Some(Value::Object(properties)) = members.get("properties") {
            if members.get("additionalProperties") == Some(&Value::Bool(false)) {
                for name in fields.keys() {
                    if !properties.contains_key(name) {
                        path.push(name.clone());
                        issues.push(ValidationIssue::new("unexpected property", path.clone()));
                        path.pop();
                    }
                }
            }
            for (name, property) in properties {
                if let Some(field) = fields.get(name) {
                    path.push(name.clone());
                    collect(property, field, path, issues);
                    path.pop();
                }
            }
        }
    }

    if let (Value::Array(entries), Some(items)) = (value, members.get("items")) {
        for (index, entry) in entries.iter().enumerate() {
            path.push(index.to_string());
            collect(items, entry, path, issues);
            path.pop();
        }
    }
}

fn matches_type(expected: &Value, value: &Value) -> bool {
    match expected {
        Value::String(name) => matches_type_name(name, value),
        Value::Array(names) => names.iter().any(|name| matches_type(name, value)),
        _ => true,
    }
}

fn matches_type_name(name: &str, value: &Value) -> bool {
    match name {
        "object" => value.is_object(),
        "array" => value.is_array(),
        "string" => value.is_string(),
        "number" => value.is_number(),
        "integer" => value.as_i64().is_some() || value.as_u64().is_some(),
        "boolean" => value.is_boolean(),
        "null" => value.is_null(),
        _ => true,
    }
}

fn type_name(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::Bool(_) => "boolean",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn add_schema() -> Value {
        serde_json::json!({
            "type": "object",
            "properties": { "a": { "type": "number" }, "b": { "type": "number" } },
            "required": ["a", "b"]
        })
    }

    #[test]
    fn a_wrong_property_type_and_a_missing_property_are_both_reported() {
        let issues = check(&add_schema(), &serde_json::json!({ "a": "not-a-number" }))
            .expect_err("the input is invalid twice over");
        assert_eq!(issues.len(), 2);
        assert!(
            issues
                .iter()
                .any(|issue| issue.path == vec!["a".to_owned()])
        );
        assert!(
            issues
                .iter()
                .any(|issue| issue.path == vec!["b".to_owned()])
        );
    }

    #[test]
    fn valid_input_passes() {
        assert!(check(&add_schema(), &serde_json::json!({ "a": 1, "b": 2 })).is_ok());
    }

    #[test]
    fn a_keyword_this_module_cannot_enforce_is_refused_up_front() {
        assert!(assert_enforceable(&add_schema()).is_ok());
        let problem = assert_enforceable(&serde_json::json!({ "oneOf": [] }))
            .expect_err("oneOf is not enforceable here");
        assert!(problem.contains("oneOf"));
    }
}

File: sdks/cpp/conformance-host/schema_subset.cpp
#include "schema_subset.hpp"

#include <algorithm>
#include <array>
#include <string_view>

namespace conformance {
namespace {

using tesseron::Json;
using tesseron::ValidationIssue;

constexpr std::array<std::string_view, 10> kSupportedKeywords = {
    "$schema", "additionalProperties", "const",    "default",  "description",
    "enum",    "items",                "properties", "required", "type"};

const char* type_name(const Json& value) {
  if (value.is_null()) return "null";
  if (value.is_boolean()) return "boolean";
  if (value.is_number()) return "number";
  if (value.is_string()) return "string";
  if (value.is_array()) return "array";
  if (value.is_object()) return "object";
  return "unknown";
}

bool matches_type_name(const std::string& name, const Json& value) {
  if (name == "object") return value.is_object();
  if (name == "array") return value.is_array();
  if (name == "string") return value.is_string();
  if (name == "number") return value.is_number();
  if (name == "integer") return value.is_number_integer();
  if (name == "boolean") return value.is_boolean();
  if (name == "null") return value.is_null();
  return true;
}

bool matches_type(const Json& expected, const Json& value) {
  if (expected.is_string()) return matches_type_name(expected.get<std::string>(), value);
  if (expected.is_array()) {
    return std::any_of(expected.begin(), expected.end(),
                       [&value](const Json& name) { return matches_type(name, value); });
  }
  return true;
}

const Json* member(const Json& object, const char* key) {
  const auto found = object.find(key);
  return found == object.end() ? nullptr : &*found;
}

void collect(const Json& schema, const Json& value, std::vector<std::string>& path,
             std::vector<ValidationIssue>& issues) {
  if (!schema.is_object()) return;

  if (const Json* const expected = member(schema, "type")) {
    if (!matches_type(*expected, value)) {
      issues.push_back(
          ValidationIssue{"expected type " + expected->dump() + ", got " + type_name(value), path});
      return;
    }
  }
  if (const Json* const allowed = member(schema, "enum");
      allowed != nullptr && allowed->is_array()) {
    if (std::find(allowed->begin(), allowed->end(), value) == allowed->end()) {
      issues.push_back(ValidationIssue{"expected one of " + allowed->dump(), path});
    }
  }
  if (const Json* const expected = member(schema, "const")) {
    if (*expected != value) issues.push_back(ValidationIssue{"expected " + expected->dump(), path});
  }

  if (value.is_object()) {
    if (const Json* const required = member(schema, "required");
        required != nullptr && required->is_array()) {
      for (const Json& name : *required) {
        if (!name.is_string()) continue;
        const auto& key = name.get_ref<const std::string&>();
        if (value.contains(key)) continue;
        path.push_back(key);
        issues.push_back(ValidationIssue{"required property is missing", path});
        path.pop_back();
      }
    }
    if (const Json* const properties = member(schema, "properties");
        properties != nullptr && properties->is_object()) {
      const Json* const additional = member(schema, "additionalProperties");
      if (additional != nullptr && additional->is_boolean() && !additional->get<bool>()) {
        for (const auto& field : value.items()) {
          if (properties->contains(field.key())) continue;
          path.push_back(field.key());
          issues.push_back(ValidationIssue{"unexpected property", path});
          path.pop_back();
        }
      }
      for (const auto& declared : properties->items()) {
        const auto field = value.find(declared.key());
        if (field == value.end()) continue;
        path.push_back(declared.key());
        collect(declared.value(), *field, path, issues);
        path.pop_back();
      }
    }
  }

  if (const Json* const items = member(schema, "items"); items != nullptr && value.is_array()) {
    for (std::size_t index = 0; index < value.size(); ++index) {
      path.push_back(std::to_string(index));
      collect(*items, value[index], path, issues);
      path.pop_back();
    }
  }
}

}  // namespace

std::optional<std::string> unenforceable_keyword(const Json& schema) {
  if (!schema.is_object()) return "a schema must be a JSON object, got " + schema.dump();
  for (const auto& entry : schema.items()) {
    const bool supported = std::find(kSupportedKeywords.begin(), kSupportedKeywords.end(),
                                     std::string_view(entry.key())) != kSupportedKeywords.end();
    if (!supported) {
      return "this adapter cannot enforce the JSON Schema keyword \"" + entry.key() + "\"";
    }
  }
  if (const Json* const properties = member(schema, "properties");
      properties != nullptr && properties->is_object()) {
    for (const auto& property : properties->items()) {
      if (auto refusal = unenforceable_keyword(property.value())) return refusal;
    }
  }
  if (const Json* const items = member(schema, "items")) {
    if (auto refusal = unenforceable_keyword(*items)) return refusal;
  }
  return std::nullopt;
}

std::vector<ValidationIssue> check(const Json& schema, const Json& value) {
  std::vector<ValidationIssue> issues;
  std::vector<std::string> path;
  collect(schema, value, path, issues);
  return issues;
}

}  // namespace conformance
